fix(lsp): lsp.buf.definition() handles responses from multiple servers

antonk52 opened the pull request:
Resolves #22318 when vim.lsp.buf.definition() subsequently jumps to multiple locations if more than one server responds.

Motivation: In case of multiple servers supporting the same lsp method, neovim should merge the results and handle them the same way it currently handles multiple results from a single server. This can reduce confusion by removing jumping to multiple locations without the user's action. In comparison vscode merges responses from its plugins/lsp servers and presents them to the user in a single picker. This is also partially solved in [lspsaga](https://nvimdev.github.io/lspsaga/definition/), where in `peek_definition` every result in opened in a separate float. 

I am uncertain whether this is the best place to handle the entire thing and open for feedback. As of opening this PR is in incomplete state as there are no tests or documentation. I did not go any further as I am not sure what would be a good way to pass multiple contexts in case the option `on_list` is set.

In the future the same approach can be applied to other methods such as `type_definition`, `declaration`, `implementation`, and the general idea of merging results from multiple servers for `hover`.

antonk52 committed:
fix(lsp): lsp.buf.definition() handles multiple responses

MariaSolOs commented on runtime/lua/vim/lsp/buf.lua:
```diff
@@ -76,8 +76,75 @@ end
+    end
+
+    -- jump to the only response
+    if total_items == 1 then
```
We should still check if `options.on_list` is given here and invoke it if so. See https://github.com/neovim/neovim/pull/25830 for context.

antonk52 replied:
Thanks, fixed in [4e72d41](https://github.com/neovim/neovim/pull/26507/commits/4e72d41ca5833f174c340d19fdefc13f4e4f91c8)

antonk52 committed:
fix(lsp): prefer on_list to jumping to only location

antonk52 committed:
fix(lsp): lsp.buf.definition() ignore nested loops

antonk52 commented:
I have thought more about the contexts for lsp results and it seems like there are a couple of potential solutions here

- Omit contexts completely, given that they are already [omitted in location_handler](https://github.com/neovim/neovim/blob/8fb7419d7c5e2df3b792d18fa56f973088e69be2/runtime/lua/vim/lsp/handlers.lua#L425). Thus this won't be a regression.
- Make each item include the context of the lsp response. The only downside I see in this is that the items are of shape `setqflist-what` (see help) and this change would somewhat break this convention.

Besides this concern, I think functionality wise this PR is a good state. I am happy to do the same for type_definition, declaration, implementation, and references in the same PR or do it separately in different ones.

mfussenegger commented on runtime/lua/vim/lsp/buf.lua:
```diff
@@ -76,8 +76,76 @@ end
+  local bufnr = api.nvim_get_current_buf()
+
+  ---@param results table<integer, {error: lsp.ResponseError, result: lsp.Location|lsp.Location[]|lsp.LocationLink[]|nil}>
+  local function handler(results)
+    local client_id_to_items = {}
```
With https://github.com/neovim/neovim/pull/26694 merged you could add a type annotation for this:

```suggestion
  local function handler(results)
    ---@type table<integer, vim.lsp.util.LocationItem[]>
    local client_id_to_items = {}
```

antonk52 commented on runtime/lua/vim/lsp/buf.lua:
```diff
@@ -76,8 +76,76 @@ end
+      return options.on_list({
+        title = title,
+        items = items,
+        -- FIXME currently passing nil as items can come from multiple clients
+        -- need a better way to attribute items to context
+        context = nil,
```
One reason having a corresponding context for each item is to understand which lsp the item is coming from. This can be a part of a custom display of multiple items and doesn't have to be supported in neovim itself. I think this makes sense similar to how you can enable display of diagnostics source.

MariaSolOs replied:
I wonder if providing the context via a separate API (similar to how you can get a client's information given its ID) would be easier/cleaner.

mfussenegger replied:
> One reason having a corresponding context for each item is to understand which lsp the item is coming from. This can be a part of a custom display of multiple items and doesn't have to be supported in neovim itself. I think this makes sense similar to how you can enable display of diagnostics source.


The current shape/structure won't let you do that because there's no association between a individual item and the context.


We'd either have to change the complete `on_list` parameter shape - which would be breaking, or add a `client_id` to each individual item.

Either way, this doesn't need to block this PR.

antonk52 committed:
fix(lsp): add type annotation

antonk52 committed:
fix(lsp): remove unnecessary check

antonk52 committed:
fix(lsp): remove context field for definition results

antonk52 committed:
fix(lsp): mention vim.lsp.buf.defintion handling multiple clients in news.txt

antonk52 commented:
I have rebased the branch and addressed the latest comments.
Please let me know if there is anything else needed from my side for this PR to be merged.

wookayin commented on runtime/lua/vim/lsp/buf.lua:
```diff
@@ -76,8 +76,74 @@ end
+  local bufnr = api.nvim_get_current_buf()
+
+  ---@param results table<integer, {error: lsp.ResponseError, result: lsp.Location|lsp.Location[]|lsp.LocationLink[]|nil}>
+  local function handler(results)
```
Hmm, the definition method now uses a separate handler; but I don't think such a duplicate is a good idea. Would it be possible to consolidate?

https://github.com/neovim/neovim/blob/master/runtime/lua/vim/lsp/handlers.lua#L450

antonk52 replied:
I agree that duplication is not optimal. As I mentioned in the PR description, ideally the same should be applied to `type_definition`, `declaration`, and `implementation` as they also can receive locations from multiple clients and all use `location_handler` currently. There are a couple of reasons why a custom handler was introduced:
- it merges and flattens potentially heterogeneous responses from multiple clients
- it has `options` in a closure to check for `on_list`
- since the responses come from multiple lsp's, it has to do extra work to use the correct client offset_encoding to turn results into location items via [`util.locations_to_items`](https://github.com/neovim/neovim/blob/master/runtime/lua/vim/lsp/handlers.lua#L430) or [`util.jump_to_locations`](https://github.com/neovim/neovim/blob/master/runtime/lua/vim/lsp/handlers.lua#L438) in case there is only single result.

Alternatively, if you agree that other location handling lsp methods should be using the same mechanism. I can update the location_handler you linked to pretty much this implementation that would handle results from multiple clients. And update `vim.lsp.buf_request_all` to accept an extra parameter `options` that would be passed to handler along with `results`.

This way the `vim.lsp.buf.defintion` and the other location methods would looks like this

```lua
function M.definition(options)
  local params = util.make_position_params()
  local bufnr = api.nvim_get_current_buf()

  vim.lsp.buf_request_all(bufnr, ms.textDocument_definition, params, location_handler, options)
end
```

If we can can have this mechanism for other lsp methods that handle locations, the current location handler(the function you linked) will be no longer needed and could be removed.

What do you think?

wookayin reviewed (commented):
I understand writing tests for LSP behaviors would be not very easy because of the need of writing mocks and the current test suite is a bit messy, but I believe this is a sort of complex behavior that could benefit a lot by writing [test cases](https://github.com/neovim/neovim/blob/master/CONTRIBUTING.md#pull-requests-prs).

antonk52 commented:
> I understand writing tests for LSP behaviors would be not very easy because of the need of writing mocks and the current test suite is a bit messy, but I believe this is a sort of complex behavior that could benefit a lot by writing [test cases](https://github.com/neovim/neovim/blob/master/CONTRIBUTING.md#pull-requests-prs).

I agree that this PR touches pretty fundamental part of LSP that users interface with and that it should be tested. Can you please link an existing test that uses mocks and tests any LSP behaviour, if there is one?

I am happy to add tests once we agree on the implementation.

antonk52 commented:
Related: Telescope added handling requests to multiple language servers in lsp_* methods 

https://github.com/nvim-telescope/telescope.nvim/commit/3b1600d

justinmk commented:
> With [#30877](https://github.com/neovim/neovim/pull/30877) definitions from multiple servers are now merged/concattenated
